pkg/proxy/proxy.go

```diff
@@ -396,6 +397,93 @@ func New(cfg *Config, events []string) *Proxy {
 	return p
 }
 
+// Init initializes a new Proxy
+// This function replaces the deprecated New function
+func Init(cfg *Config) (*Proxy, error) {
+	if cfg.Log == nil {
+		cfg.Log = &log.Logger{Out: ioutil.Discard}
+	}
+
+	// if no events are passed, listen for all events
+	if len(cfg.Events) == 0 {
+		cfg.Events = []string{"*"}
+	} else {
+		for _, event := range cfg.Events {
+			if _, found := validEvents[event]; !found {
+				fmt.Printf("Warning: You're attempting to listen for \"%s\", which isn't a valid event\n", event)
+			}
+		}
+	}
+
+	// build endpoint routes
+	var endpointRoutes []EndpointRoute
+	if cfg.UseConfiguredWebhooks {
+		// build from user's API config
+		endpoints := getEndpointsFromAPI(cfg.Key, cfg.APIBaseURL)
+		if len(endpoints.Data) == 0 {
+			// TODO: this log needs to be handled differently
+			return nil, errors.New("You have not defined any webhook endpoints on your account. Go to the Stripe Dashboard to add some: https://dashboard.stripe.com/test/webhooks")
+		}
+		endpointRoutes = buildEndpointRoutes(endpoints, parseURL(cfg.ForwardURL), parseURL(cfg.ForwardConnectURL), cfg.ForwardHeaders, cfg.ForwardConnectHeaders)
+	} else {
+		// build from --forward-to urls
+		if len(cfg.ForwardConnectURL) == 0 {
+			cfg.ForwardConnectURL = cfg.ForwardURL
+		}
+		if len(cfg.ForwardConnectHeaders) == 0 {
+			cfg.ForwardConnectHeaders = cfg.ForwardHeaders
+		}
+		// non-connect endpoints
+		endpointRoutes = append(endpointRoutes, EndpointRoute{
+			URL:            parseURL(cfg.ForwardURL),
+			ForwardHeaders: cfg.ForwardHeaders,
+			Connect:        false,
+			EventTypes:     cfg.Events,
+		})
+		// connect endpoints
+		endpointRoutes = append(endpointRoutes, EndpointRoute{
+			URL:            parseURL(cfg.ForwardConnectURL),
+			ForwardHeaders: cfg.ForwardConnectHeaders,
+			Connect:        true,
+			EventTypes:     cfg.Events,
+		})
+	}
+
+	p := &Proxy{
+		cfg: cfg,
+		stripeAuthClient: stripeauth.NewClient(cfg.Key, &stripeauth.Config{
+			Log:        cfg.Log,
+			APIBaseURL: cfg.APIBaseURL,
+		}),
+		events: convertToMap(cfg.Events),
+	}
+
+	for _, route := range endpointRoutes {
+		// append to endpointClients
+		p.endpointClients = append(p.endpointClients, NewEndpointClient(
+			route.URL,
+			route.ForwardHeaders,
+			route.Connect,
+			route.EventTypes,
+			&EndpointConfig{
+				HTTPClient: &http.Client{
+					CheckRedirect: func(req *http.Request, via []*http.Request) error {
+						return http.ErrUseLastResponse
+					},
+					Timeout: defaultTimeout,
+					Transport: &http.Transport{
+						TLSClientConfig: &tls.Config{InsecureSkipVerify: cfg.SkipVerify},
+					},
+				},
+				Log:             p.cfg.Log,
+				ResponseHandler: EndpointResponseHandlerFunc(p.processEndpointResponse),
```

Comment:
To stream events with gRPC, I would want to have a different ResponseHandler. Maybe a field of `ResponseHandler EndpointResponseHandlerFunc` could be added to the `Config` struct?

Reply:
Yeah we need to update that. We need to be able to both continue to send the webhook response to stripe and at the same time stream that resp to gRPC.

Inside `processEndpointResponse` we already check `if p.webSocketClient != nil {` to send the msg through the websocket, so maybe just adding the gRPC stream handler on `p` so we can also check `if p.gRPCService` or something?

There is also only 1 `EndpointResponseHandlerFunc` type used in the CLI, and it's this occurrence, so I'm not sure why we have it dynamic this way. There's probably a better way of doing this.

Reply:
I'm ok with having something like `if p.gRPCService ` inside of `processEndpointResponse`. If we go with this approach, then I can add that field when I implement the gRPC method. Then this LGTM.

Reply:
Seems good enough for me too.